engine/engine.go

```diff
@@ -186,9 +186,12 @@ func (e *EngineAPI) ForkchoiceUpdatedV3(
 	//   It must return STATUS_VALID if all of the transactions could be executed without error.
 	// TODO checktx doesn't actually run the tx, it only does basic validation.
 	for _, txBytes := range cosmosTxs {
-		if _, err := e.txValidator.CheckTx(context.Background(), &abci.RequestCheckTx{
+		if checkTxResult, err := e.txValidator.CheckTx(context.Background(), &abci.RequestCheckTx{
 			Tx: txBytes,
-		}); err != nil {
+		}); err != nil || checkTxResult.IsErr() {
+			// TODO what is the difference between err != nil and checkTxResult.IsErr()?
```

Comment:
I took a look at how to [cosmos sdk baseapp](https://github.com/cosmos/cosmos-sdk/blob/339e26ea8f3a1f2c81ae2fa8064a73f24dd96883/baseapp/abci.go#L363) implements `CheckTx` and `err` is only returned at the top level if a request type other than `CheckTxType_New` or `CheckTxType_Recheck` is passed through. Otherwise, relevant error information from validating the transaction will be stored in the response struct.

Reply:
Thank you :raised_hands: 

I'll update the code based on that information and merge.